pinctrl: add STM32 implementation

erwango opened the pull request:
Based on https://github.com/zephyrproject-rtos/zephyr/pull/39430 (Thanks @gmarull for the great initial work)

Following merge of pinctrl API (#37572), provide the STM32 implementation.
This takes place of the existing, similar but non generic STM32 implementation.

All in-tree drivers are converted, along with the peripheral configuration for each in tree boards.

Deprecation of the previous API is not addressed, to comply with deprecation timeline proposed in https://github.com/zephyrproject-rtos/zephyr/issues/39740

Replaces https://github.com/zephyrproject-rtos/zephyr/pull/39430

Important notice:
`pinctrl-0` and `pinctrl-names` are now set as required properties (for pinctrl user devices).
Main reason is to avoid DT_API cryptic error messages when not provided.

Note: Set as DNM as it includes a BK temp patch to increase workload capacity

EDIT: Compliance check reports a typdef warning, which I suggest to discard.
```
-:9321: WARNING:NEW_TYPEDEFS: do not add new typedefs
#9321: FILE: soc/arm/st_stm32/common/pinctrl_soc.h:32:
+typedef struct pinctrl_soc_pin {
```

EDIT: 11/15: ~~Removed PWM driver update, due to underlying issue in PWM driver that should be taken care first and separately. Submitted https://github.com/zephyrproject-rtos/zephyr/pull/40430 to handle PWM case.~~
EDIT: 11/23: Based on #40515 to remove need to use `PINCTRL_STORE_REG` on F1 series. As a consequence, PWM driver update is back.
EDIT: 11/24: Added a couple of commits to fix erroneous board descriptions that are now leading to build issues (now that pinctrl-0 and pinctrl-names are required)

gmarull committed:
drivers: pinctrl: stm32: initial version

Add initial version for STM32 pinctrl driver. Driver has been written
re-using many of the already existing parts in
drivers/pinmux/pinmux_stm32.c.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: arm: stm32: add pinctrl state name for UART peripheral

Add the pinctrl state name (default) for the UART/USART/LPUART
peripherals. Changes performed using the following Python script run
from the repository root:

```
from pathlib import Path
import re

for fpath in Path(".").glob("boards/arm/**/*.dts"):
    lines = open(fpath).readlines()

    is_stm32 = False
    for line in lines:
        if "stm32" in line:
            is_stm32 = True
            break

    if not is_stm32:
        continue

    with open(fpath, "w") as f:
        for line in lines:
            f.write(line)

            m = re.match(r"(\s+)pinctrl-0.*us?art.*", line)
            if m:
                f.write(m.group(1) + "pinctrl-names = \"default\";\n")
```

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
drivers: serial: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
boards: arm: stm32: enable pinctrl driver

Enable the pin control driver on all STM32 based boards. The following
script has been used to do this task:

```
from pathlib import Path
import re

for fpath in Path(".").glob("boards/arm/**/*_defconfig"):
    lines = open(fpath).readlines()

    is_stm32 = False
    for line in lines:
        if "CONFIG_SOC_SERIES_STM32" in line:
            is_stm32 = True
            break

    if not is_stm32:
        continue

    lines += ["\n", "# enable pin controller\n", "CONFIG_PINCTRL=y\n"]

    with open(fpath, "w") as f:
        f.writelines(lines)
```

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

erwango committed:
drivers/adc: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for ADC peripheral

Add the pinctrl state name (default) for the ADC peripherals.
Changes performed based on the script proposed in
"boards: arm: stm32: add pinctrl state name for UART peripheral"

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/can: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for CAN peripheral

Add the pinctrl state name (default) for the CAN peripherals.
Changes performed based on the script proposed in
"boards: arm: stm32: add pinctrl state name for UART peripheral"

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/fdcan: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for FDCAN peripheral

Add the pinctrl state name (default) for the FDCAN peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/dac: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for DAC peripheral

Add the pinctrl state name (default) for the CAN peripherals.
Changes performed based on the script proposed in
"boards: arm: stm32: add pinctrl state name for UART peripheral"

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/disk: sdmmc: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for SDMMC peripheral

Add the pinctrl state name (default) for the SDMMC peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/ethernet: stm32_hal: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for ethernet peripheral

Add the pinctrl state name (default) for the ethernet peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/ethernet: dwmac_stm32h7x: use new pinctrl API

Use the new pinctrl API to configure pins.

Additionally, rename eth0_pins to eth0_pcfg to better fit
new pinctrl API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/flash: qspi stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for QSPI peripheral

Add the pinctrl state name (default) for the QSPI peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: nucleo_f446ze: Remove QSPI node definition

There's no QSPI peripheral device available on this board.
Remove node definition.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for I2C peripheral

Add the pinctrl state name (default) for the I2C peripherals.
Changes performed based on the script proposed in
"boards: arm: stm32: add pinctrl state name for UART peripheral"

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/i2c: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for I2S peripheral

Add the pinctrl state name (default) for the I2S peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/i2s: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/memc: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for FMC peripheral

Add the pinctrl state name (default) for the FMC peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/pwm: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.
Since STM32F1 series require pinctrl option and required register
address is parent timer address in place of own node register address,
use PINCTRL_DT_INST_CUSTOM_REG_DEFINE in place of usual
PINCTRL_DT_INST_DEFINE for this specific series.

Additionally, remove the automatic selection of PINMUX API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for PWM peripheral

Add the pinctrl state name (default) for the PWM peripherals.
Changes performed based on the script proposed in
"boards: arm: stm32: add pinctrl state name for UART peripheral"

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/spi: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for SPI peripheral

Add the pinctrl state name (default) for the CAN peripherals.
Changes performed based on the script proposed in
"boards: arm: stm32: add pinctrl state name for UART peripheral"

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/usb/device: stm32: use new pinctrl API

Use the new pinctrl API to configure pins.

Additionally, rename usb_pinctrl to usb_pcfg to better fit
new pinctrl API.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: arm: stm32: add pinctrl state name for USB peripheral

Add the pinctrl state name (default) for the USB peripherals.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32: Remove use of CONFIG_PINMUX

Now that all drivers and all boards have been converted to the
use of PINCTRL, remove usage of PINMUX.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

gmarull commented on drivers/ethernet/eth_dwmac_stm32h7x.c:
```diff
@@ -57,8 +58,7 @@ int dwmac_bus_init(struct dwmac_priv *p)
 
-	ret = stm32_dt_pinctrl_configure(eth0_pins, ARRAY_SIZE(eth0_pins),
-					 (uint32_t)p->base_addr);
+	ret = pinctrl_apply_state(eth0_pins, PINCTRL_STATE_DEFAULT);
```
nit: probably adjust the variable name here, `pcfg`?

erwango replied:
Makes sense. Will update

gmarull commented on drivers/usb/device/usb_dc_stm32.c:
```diff
@@ -60,9 +60,10 @@ LOG_MODULE_REGISTER(usb_dc_stm32);
-						ST_STM32_DT_INST_PINCTRL(0, 0);
 
+PINCTRL_DT_INST_DEFINE(0)
+static const struct soc_gpio_pinctrl *usb_pinctrl =
```
ditto

erwango replied:
Makes sense. Will update

gmarull reviewed (approved):
LGTM, a couple of minor comments. Another suggestion: drivers now depend on `CONFIG_PINCTRL=y`, so should we specify such dependency in Kconfig?

erwango commented:
@gmarull Thanks for the review

> Another suggestion: drivers now depend on CONFIG_PINCTRL=y, so should we specify such dependency in Kconfig?

Most drivers also depends on CLOCK_CONTROL and this dependency is not specified.
So, for coherency, if we start stating these dependencies, we should perform a full assessment and update.

gmarull commented:
Another suggestion: may be useful to add a test for DT parsing: https://github.com/zephyrproject-rtos/zephyr/pull/39429/commits/50bd669f6bc8329fdd51c315c971ef0dea4ed64e

erwango commented:
@gmarull Fyi, I'm facing an issue with PWM STM32 driver on F1 (which uses `CONFIG_PINCTRL_STORE_REG=y`) devices where `Z_PINCTRL_DEV_CONFIG_INIT` won't work as for this device we need to provide parent reg address instead of device one.
I've worked around this creating a `PINCTRL_DT_INST_PARENT_DEFINE` which ends up calling a `Z_PINCTRL_DEV_PARENT_CONFIG_INIT` that performs `.reg = DT_REG_ADDR(DT_PARENT(node_id)),`, but I'm not convinced about the macro name. Suggestion welcome about naming (or alternative, more elegant solution).

gmarull commented:
> @gmarull Fyi, I'm facing an issue with PWM STM32 driver on F1 (which uses `CONFIG_PINCTRL_STORE_REG=y`) devices where `Z_PINCTRL_DEV_CONFIG_INIT` won't work since, for this device, we need to provide parent reg address instead of device's one. I've worked around this creating a `PINCTRL_DT_INST_PARENT_DEFINE` which ends up calling a `Z_PINCTRL_DEV_PARENT_CONFIG_INIT` that performs `.reg = DT_REG_ADDR(DT_PARENT(node_id)),`, but I'm not convinced about the macro name. Suggestion welcome about naming (or alternative, more elegant solution).

I wonder if the `pinctrl-n` entries should be moved to the timer node instead, since it's really a timer level configuration (PWM is not a device on its own in case of STM32). Similar to what is being done for the prescaler setting.

MaureenHelm reviewed (approved):
nice work!

ABOSTM reviewed (changes requested):
Few remarks with regards to this huge work.

erwango commented on drivers/pinctrl/Kconfig.stm32:
```diff
@@ -0,0 +1,17 @@
+	help
+	  Enable pin controller driver for STM32 MCUs
+
+config PINCTRL_STM32_REMAP_INIT_PRIORITY
+	int "Remap initialization priority"
+	default 2
+	help
+	  Initialization priority for the routine in charge of configuring the
+	  remap for pins PA11/12.
```
Copy/paste of initial code in `drivers/pinmux/Kconfig.stm32`, I propose to address this in a different PR if required.

erwango commented on drivers/can/can_stm32fd.c:
```diff
@@ -258,12 +259,15 @@ static const struct can_stm32fd_config can_stm32fd_cfg_##inst = {              \
 #else /* CONFIG_CAN_FD_MODE */
 
 #define CAN_STM32FD_CFG_INST(inst)                                             \
+									       \
+PINCTRL_DT_INST_DEFINE(inst)						       \
+									       \
```
Indentation is aligned on the other parts of the macro

erwango commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7_defconfig:
```diff
@@ -31,3 +31,7 @@ CONFIG_UART_CONSOLE=y
 # enable pin controller
 CONFIG_PINCTRL=y
+
+
+CONFIG_ETH_DWMAC=y
+CONFIG_ETH_STM32_HAL=n
```
Good catch, thanks

erwango commented:
> I wonder if the `pinctrl-n` entries should be moved to the timer node instead, since it's really a timer level configuration: we are routing timer signals (PWM is not a device on its own in case of STM32). Similar to what is being done for the prescaler setting.

@gmarull 

Note:  I've removed PWM driver update from this series as I'd like to be able to progress on this PR. If PWM driver is used PINMUX driver will be enabled. That's not ideal, but only temporary.

Your proposal makes sense in a way but differs from Linux bindings (cf example here: https://elixir.bootlin.com/linux/v5.16-rc1/source/arch/arm/boot/dts/stm32mp15xx-dkx.dtsi#L569). Explanation from the Linux part is that pwm is part of a multi function driver and pin cfg is handled on the pwm part as it wouldn't make much sense to handle on the common part. This makes sense as well.

Are we sure this would be the only case where this configuration is faced ?

gmarull commented:
> > I wonder if the `pinctrl-n` entries should be moved to the timer node instead, since it's really a timer level configuration: we are routing timer signals (PWM is not a device on its own in case of STM32). Similar to what is being done for the prescaler setting.
> 
> @gmarull
> 
> Note: I've removed PWM driver update from this series as I'd like to be able to progress on this PR. If PWM driver is used PINMUX driver will be enabled. That's not ideal, but only temporary.
> 
> Your proposal makes sense in a way (and is elegant) but differs from Linux bindings (cf example here: [elixir.bootlin.com/linux/v5.16-rc1/source/arch/arm/boot/dts/stm32mp15xx-dkx.dtsi#L569](https://elixir.bootlin.com/linux/v5.16-rc1/source/arch/arm/boot/dts/stm32mp15xx-dkx.dtsi#L569)). Explanation from the Linux part is that pwm is part of a multi function driver and pin cfg is handled on the pwm part as it wouldn't make much sense to handle on the common part. This makes sense as well.
> 
> Are we sure this would be the only case where this configuration is faced ?

The same could be said for `st,prescaler` setting, now part of the timer node but always used by children (and it can be set to just one value, not two). A timer will likely be used by just one child, or at least only one of the children will use timer signals (timers just have `CH1..4[N]` and `BKIN[1-2]`). Maybe I'm wrong, but I think it should always be the case.

Another question: have you considered the option to encode remap information in the generated nodes? It is something possible, and it allows driver to stop requiring `CONFIG_PINCTRL_STORE_REG`. See https://github.com/zephyrproject-rtos/zephyr/pull/39974/commits/3964d424bf229af2c8bd44bf8ddd09364f13515d#diff-ee769543e50316a2c7f8bc833266a51c8253e6084e3b240fb9b75a889e2b5f4eR176 https://github.com/zephyrproject-rtos/hal_gigadevice/blob/main/include/dt-bindings/pinctrl/gd32f403z(k-i-g-e-c-b)xx-pinctrl.h#L229. It would just require re-generating F1 nodes and pinctrl driver, but users would not even notice.

erwango commented:
> The same could be said for `st,prescaler` setting, now part of the timer node but always used by children (and it can be set to just one value, not two). A timer will likely be used by just one child, or at least only one of the children will use timer signals (timers just have `CH1..4[N]` and `BKIN[1-2]`). Maybe I'm wrong, but I think it should always be the case.

True, but this one doesn't exist in Linux binding.

> Another question: have you considered the option to encode remap information in the generated nodes?

Looks interesting indeed. The question I have is around generation, were we'll miss details on AFIO registers in source .xml files.
So we kinda need to encode this beforehand and store in an additional dedicated generation input file. Am I missing something?

gmarull commented:
> > Another question: have you considered the option to encode remap information in the generated nodes?
> 
> Looks interesting indeed. The question I have is around generation, were we'll miss details on AFIO registers in source .xml files. So we kinda need to encode this beforehand and store in an additional dedicated generation input file. Am I missing something?

You'll need something like this: https://github.com/zephyrproject-rtos/hal_gigadevice/blob/main/include/dt-bindings/pinctrl/gd32f403xx-afio.h. Then based on pin availability and signal names (information XML) you should have everything necessary. STM32 case should not be too complex since only F1 uses this pinmux model. MAPR registers seem to be equal amongst F1 SoCs, but some have extra fields:

![image](https://user-images.githubusercontent.com/25011557/142405861-4d4931c9-9dd9-481e-8d92-265c61800e53.png)

erwango commented:
> You'll need something like this: https://github.com/zephyrproject-rtos/hal_gigadevice/blob/main/include/dt-bindings/pinctrl/gd32f403xx-afio.h. 

Ok, so aside from this tedious manual definition part, this looks globally a reasonable effort. Will give it a try. Thanks!

erwango commented on dts/bindings/serial/st,stm32-uart.yaml:
```diff
@@ -2,21 +2,11 @@ description: STM32 UART
     interrupts:
       required: true
-
-    pinctrl-0:
```
To avoid frustration on build errors, the following should be added in all impacted bindings.
```
    pinctrl-names:
      required: true
```

gmarull replied:
I initially planned to make `pinctrl-0` and `pinctrl-names` required, but then dropped the idea because Nordic needs to maintain backwards compatibility. Maybe nrf bindings can set required false?

erwango replied:
If possible, it would indeed make sense. But since we're doing this for interrupts and reg already, I guess we can live with the same on pinctrl-names...
But indeed, if it is not set as required in common binding, this is something we should take care when reviewing the related drivers changes.

mbolivar-nordic replied:
> Maybe nrf bindings can set required false?

You can't "downgrade" a `required: true` to a `required: false`:

> Note that weakening requirements by having required: false where the included file has required: true is an error. This is meant to keep the organization clean.

https://docs.zephyrproject.org/latest/guides/dts/bindings.html#include

mbolivar-nordic replied:
>  if it is not set as required in common binding, this is something we should take care when reviewing the related drivers changes.

so I think this is the way to go

tejlmand reviewed (commented):
Commit https://github.com/zephyrproject-rtos/zephyr/pull/40194/commits/a12abce057a2588637ff0acc06ea4d003424333b breaks bisectability:
```
$ git checkout 
$ cmake -GNinja -DBOARD=stm32h747i_disco_m7 -Bbuild samples/hello_world/
....
$ ninja -Cbuild
FAILED: zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj 
ccache /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc -DBUILD_VERSION=v2.7.99-1489-ga12abce057a2 -DCORE_CM7 -DHSE_VALUE=25000000 -DKERNEL -DSTM32H747xx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../include -Izephyr/include/generated -I../soc/arm/st_stm32/stm32h7 -I../drivers -I../soc/arm/st_stm32/common -I/projects/github/ncs/modules/hal/cmsis/CMSIS/Core/Include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/soc -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include/Legacy -I/projects/github/ncs/modules/hal/stm32/stm32cube/common_ll/include -isystem ../lib/libc/minimal/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include-fixed -Os -imacros /projects/github/ncs/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m7 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/projects/github/ncs/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/projects/github/ncs/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/projects/github/ncs=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj -MF zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj.d -o zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj -c /projects/github/ncs/zephyr/drivers/serial/uart_stm32.c
In file included from ../include/sys/util_macro.h:34,
                 from ../include/sys/atomic.h:16,
                 from ../include/kernel_includes.h:21,
                 from ../include/kernel.h:17,
                 from /projects/github/ncs/zephyr/drivers/serial/uart_stm32.c:17:
../include/drivers/pinctrl.h:89:10: error: 'PINCTRL_STATE_DT_N_S_soc_S_serial_40011000_PINCTRL_IDX_0_UPPER_TOKEN' undeclared here (not in a function)
   89 |  _CONCAT(PINCTRL_STATE_,             \
      |          ^~~~~~~~~~~~~~
../include/sys/util_internal.h:67:26: note: in definition of macro '__DEBRACKET'
   67 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
../include/sys/util_internal.h:59:2: note: in expansion of macro '__GET_ARG2_DEBRACKET'
   59 |  __GET_ARG2_DEBRACKET(one_or_two_args _if_code, _else_code)
      |  ^~~~~~~~~~~~~~~~~~~~
../include/sys/util_internal.h:54:2: note: in expansion of macro '__COND_CODE'
   54 |  __COND_CODE(_XXXX##_flag, _if_1_code, _else_code)
      |  ^~~~~~~~~~~
...
```

Actually, build is broken already from here: 7a42cde30cc99186e19e1991d54a21d10a94ee60
but with another failure:
```
AILED: zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj 
ccache /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/arm-none-eabi-gcc -DBUILD_VERSION=v2.7.99-1478-g7a42cde30cc9 -DCORE_CM7 -DHSE_VALUE=25000000 -DKERNEL -DSTM32H747xx -DUSE_FULL_LL_DRIVER -DUSE_HAL_DRIVER -D_FORTIFY_SOURCE=2 -D__PROGRAM_START -D__ZEPHYR_SUPERVISOR__ -D__ZEPHYR__=1 -I../include -Izephyr/include/generated -I../soc/arm/st_stm32/stm32h7 -I../drivers -I../soc/arm/st_stm32/common -I/projects/github/ncs/modules/hal/cmsis/CMSIS/Core/Include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/soc -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include -I/projects/github/ncs/modules/hal/stm32/stm32cube/stm32h7xx/drivers/include/Legacy -I/projects/github/ncs/modules/hal/stm32/stm32cube/common_ll/include -isystem ../lib/libc/minimal/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include -isystem /opt/gcc-arm-none-eabi-9-2020-q2-update/bin/../lib/gcc/arm-none-eabi/9.3.1/include-fixed -Os -imacros /projects/github/ncs/zephyr/build/zephyr/include/generated/autoconf.h -ffreestanding -fno-common -g -gdwarf-4 -fdiagnostics-color=always -mcpu=cortex-m7 -mthumb -mabi=aapcs -mfp16-format=ieee -imacros /projects/github/ncs/zephyr/include/toolchain/zephyr_stdint.h -Wall -Wformat -Wformat-security -Wno-format-zero-length -Wno-main -Wno-pointer-sign -Wpointer-arith -Wexpansion-to-defined -Wno-address-of-packed-member -Wno-unused-but-set-variable -Werror=implicit-int -fno-asynchronous-unwind-tables -fno-pie -fno-pic -fno-strict-overflow -fno-reorder-functions -fno-defer-pop -fmacro-prefix-map=/projects/github/ncs/zephyr/samples/hello_world=CMAKE_SOURCE_DIR -fmacro-prefix-map=/projects/github/ncs/zephyr=ZEPHYR_BASE -fmacro-prefix-map=/projects/github/ncs=WEST_TOPDIR -ffunction-sections -fdata-sections -std=c99 -nostdinc -MD -MT zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj -MF zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj.d -o zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_stm32.c.obj -c /projects/github/ncs/zephyr/drivers/serial/uart_stm32.c
In file included from ../include/sys/util_macro.h:34,
                 from ../include/sys/atomic.h:16,
                 from ../include/kernel_includes.h:21,
                 from ../include/kernel.h:17,
                 from /projects/github/ncs/zephyr/drivers/serial/uart_stm32.c:17:
../include/drivers/pinctrl.h:89:10: error: 'PINCTRL_STATE_DT_N_S_soc_S_serial_40011000_PINCTRL_IDX_0_UPPER_TOKEN' undeclared here (not in a function)
   89 |  _CONCAT(PINCTRL_STATE_,             \
      |          ^~~~~~~~~~~~~~
../include/sys/util_internal.h:67:26: note: in definition of macro '__DEBRACKET'
   67 | #define __DEBRACKET(...) __VA_ARGS__
      |                          ^~~~~~~~~~~
```

tejlmand reviewed (approved):
Build system related changes changes looks good.

Would like someone else to comment on whether it can be accepted that commits in the PR breaks bisectability for stm32 boards.

erwango commented:
> Would like someone else to comment on whether it can be accepted that commits in the PR breaks bisectability for stm32 boards.

I think I just need to reverse commit orders (put board before drivers)

carlescufi commented:
@erwango waiting for you to rework so I can merge.

erwango committed:
dts/bindings: stm32: Set pinctrl-[0/names] properties as required

Now that STM32 drivers are using pinctrl API, set pintrl-0 and
pintrl-names properties as required in order to report malformed
nodes description soon at build stage and avoid cryptic
DT api build error messages.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
Pushed last version:
- based on #40515
- with a care taken to preserve bisectability.

FRASTM reviewed (commented):
LGTM

FRASTM commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -0,0 +1,236 @@
+	volatile uint32_t *reg;
+	uint16_t remap;
+
+	remap = (uint8_t)STM32_DT_PINMUX_REMAP(pins[0].pinmux);
```
why casting uint16_t remap to uin8_t

FRASTM commented on drivers/pinctrl/pinctrl_stm32.c:
```diff
@@ -0,0 +1,236 @@
+#endif /* (REMAP_PA11 || REMAP_PA12) || REMAP_PA11_PA12 */
+
+	return 0;
+}
```
is the returned code of the function tested elsewhere ?

erwango committed:
dts/arm/st: mp1: Default node status should be "disabled"

Set SPI nodes status as disabled, as this should be the in .dtsi
soc description.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32l5: Refactor serial peripheral descriptions

On stm32l5 based boards, some serial devices descriptions were
scattered between common.dtsi and base or non secure dts files.

This leads to build issues now that pinctrl-0 and pinctrl-names
properties are required but this was anyway broken since the
beginning.

Group all serial devices definition in common.dtsi so complete devices
description is available to both non secure and base descriptions.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
boards: stm32: Fix incomplete nodes descriptions

Now that pinctrl-0 property is required in descriptions of nodes
that implies definition of pins, these incomplete nodes definitions
lead to build error.
Fix or remove depending of the board documentation.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

ABOSTM reviewed (approved):
LGTM

erwango committed:
samples/drivers: led_pca9633: Replace target stm32373c_eval

Pinctrl is missing in i2c1 device description on stm32373c_eval
ovderlay file, besides using I2C1 this board (as mentioned in readme)
requires to remove 2 resistors on the board.

Make it simple and instead use a board with arduino_i2c port
available.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@carlescufi Aside from the compliance issue about the typdedef addition that I can't avoid, PR is good to merge.
